pwnlib/encoders/i386/ascii_shellcode.py

```diff
@@ -0,0 +1,300 @@
+        r""" Allocate enough space on the stack for the shellcode
+
+        int_size is taken from the context (context.bits / 8)
+
+        Args:
+            size (int): The allocation size
+            vocab (bytes): Allowed characters
+
+        Returns:
+            bytes: The allocator shellcode
+
+        Examples:
+
+            >>> context.update(arch='i386', os='linux')
+            >>> vocab = b'!"#$%&\'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\]^_`abcdefghijklmnopqrstuvwxyz{|}~'
+            >>> encoders.i386.ascii_shellcode.encode._get_allocator(300, vocab)
+            b'TX-!!!!-!_``-t~~~P\\%!!!!%@@@@'
+        """
+        size += 0x1e # add typical allocator size
+        int_size = context.bits // 8
+        # Use eax for subtractions because sub esp, X doesn't assemble to ascii
+        result = b'TX'  # push esp; pop eax
+        # Set target to the `size` arg
+        target = struct.pack('=I', size)
+        # All we are doing here is adding (subtracting) `size` 
+        # to esp (to allocate space on the stack), so we don't care
+        # about esp's actual value. That's why the `last` parameter 
+        # for `calc_subtractions` can just be zero
+        for subtraction in self._calc_subtractions(
+            b'\x00'*int_size, target, vocab):
+            if six.PY2:
+                subtraction = str(subtraction)
+            # sub eax, subtraction
+            result += struct.pack('=c{}s'.format(int_size), b'-', subtraction)
+        result += b'P\\'  # push eax, pop esp
+        # Zero out eax for the unpacking part
+        pos, neg = self._find_negatives(vocab)
+        # and eax, pos; and eax, neg ; (0b00010101 & 0b00101010 = 0b0)
+        result += struct.pack('=cIcI', b'%', pos, b'%', neg)
+        return result
+
+
+    @LocalContext
+    def _find_negatives(self, vocab):
+        r""" Find two bitwise negatives in the vocab so that when they are and-ed the result is 0.
+
+        int_size is taken from the context (context.bits / 8)
+
+        Args:
+            vocab (bytes): Allowed characters
+
+        Returns:
+            Tuple[int, int]: value A, value B
+
+        Raises:
+            ArithmeticError: The allowed character set does not contain
+            two characters that when they are bitwise-anded with eachother
+            they result is 0
+
+        Examples:
+
+            >>> context.update(arch='i386', os='linux')
+            >>> vocab = b'!"#$%&\'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\]^_`abcdefghijklmnopqrstuvwxyz{|}~'
+            >>> a, b = encoders.i386.ascii_shellcode.encode._find_negatives(vocab)
+            >>> a & b
+            0
+        """
+        int_size = context.bits // 8
+        for products in product(*[vocab for _ in range(2)]):
+            if six.PY3:
+                if products[0] & products[1] == 0:
+                    return tuple(int.from_bytes(x.to_bytes(1, 'little')*int_size, 'little') for x in products)

```

Comment:
would be easier to just do some
```suggestion
                    return tuple(unpack(p8(x)*int_size) for x in bytearray(products))
```
or
```suggestion
                    MAGIC = (1 << context.bits) // 255  # 0x01010101 when 32 bits
                    return tuple(x * MAGIC for x in bytearray(products))
```
which would work for both py2 and py3 (maybe omit `bytearray(` if products are guaranteed to yield ints)